STM32F405RGTx/Core/Inc/state_machine.hpp

```diff
@@ -0,0 +1,104 @@
+/* state_machine.h
+ *
+ *    Created on: Jul. 11, 2021
+ *            Author: Kanishk Kashyap
+ */
+#pragma once
+
+#include "can.h"
+#include "cmsis_os.h"
+#include "can.h"
+#include "util.hpp"
+
+#ifndef _STATE_MACHINE_H_
+#define _STATE_MACHINE_H_
+
+// Constants - These numbers are arbritrary should be changed after testing
+// Mosfet Array Length
+#define NUM_MOSFETS 6
+#define NUM_PHASE_OUTPUTS 3
+
+// Severe Fault
+#define MAX_MOSFET_TEMP_SEVERE 70.0
+#define MAX_DCVOLTAGE_SEVERE 4.0
+#define MIN_DCVOLTAGE_SEVERE 1.8
+#define MAX_VOLTAGE_SEVERE 4.0
+#define MIN_VOLTAGE_SEVERE 1.0
+#define MAX_TEMP_SEVERE 70.0
+#define MAX_CURRENT_SEVERE 50.0
+
+// Normal Fault
+#define MAX_MOSFET_TEMP_NORMAL 60.0
+#define MAX_DCVOLTAGE_NORMAL 4.0
+#define MIN_DCVOLTAGE_NORMAL 1.8
+#define MAX_VOLTAGE_NORMAL 3.8
+#define MIN_VOLTAGE_NORMAL 1.8
+#define MAX_TEMP_NORMAL 60.0
+// #define MIN_CURRENT_NORMAL 5.0
+#define MAX_CURRENT_NORMAL 40.0
+
+#define MIN_OVERVOLT_FAULTS 5
+#define MIN_UNDERVOLT_FAULTS 5
+#define MIN_TEMP_FAULTS 5
+
+// DC_CAPTEMP
+#define MAX_DCCAP_TEMP_NORMAL 5
+#define MAX_DCCAP_TEMP_SEVERE 6
+
+// TRACK_INFORMATION
+#define TRACK_LENGTH 100
+#define DISTANCE_THRESHOLD 6
+
+using namespace std;
+
+typedef enum
+{
+    Initialize,
+    InitializeFault,
+    Idle,
+    Brake,
+    AutoPilot,
+    ManualControl,
+    EmergencyBrake,
+    NormalDangerFault,
+    SevereDangerFault,
+    NoFault,
+} State_t;
+
+typedef State_t (*pfEvent)(void);
```

Comment:
After testing the BMS software on the actual hardware, I've determined that this typedef causes a runtime hard fault that's pretty hard to catch.

Refer to the changes I made in PR #20 of the BMS to see to fix this: https://github.com/waterloop/BMS_SW_G5/pull/20

Reply:
nvm, I saw that you already changed it to a switch statement in the mainloop. Just remove this typedef then if its not being used.

Reply:
@GuyBlumenthal Given the current implementation, which I followed similar to the BMS one I cannot remove this typedef as I need it to make the StateMachine struct, do you have an idea on how to approach this?

![Screen Shot 2022-04-01 at 6 15 40 PM](https://user-images.githubusercontent.com/64444910/161348160-7f141d32-5be9-422f-ac66-c3c465c9dfa1.png)